### PR TITLE
update msa and dynamic toll functions

### PR DESCRIPTION
**Pull request by ychtw**

### CODE REVIEW

File: tm2py/components/network/highway/highway_assign.py [hunk 1]
```diff
@@ -126,14 +128,20 @@ def validate_inputs(self):
     def run(self):
         """Run highway assignment."""
         demand = PrepareHighwayDemand(self.controller)
-        if self.controller.iteration >= 0:
-            demand.run()
+        iteration = self.controller.iteration
+        dst_veh_groups = self.config.tolls.dst_vehicle_group_names
+        run_dynamic_toll = self.controller.config.highway.tolls.run_dynamic_toll
```
Comment (i-am-sijia) [inline tm2py/components/network/highway/highway_assign.py]:
```suggestion
        run_dynamic_toll = self.config.tolls.run_dynamic_toll
```

File: tm2py/components/network/highway/highway_assign.py [hunk 2]
```diff
@@ -126,14 +128,20 @@ def validate_inputs(self):
...
+        valuetoll_start_tollbooth_code = self.config.tolls.valuetoll_start_tollbooth_code
+        max_dynamic_valuetoll = self.config.tolls.max_dynamic_valuetoll
+        apply_msa = self.controller.config.highway.msa.apply_msa
```
Comment (i-am-sijia) [inline tm2py/components/network/highway/highway_assign.py]:
```suggestion
        apply_msa = self.config.msa.apply_msa
```

File: tm2py/components/network/highway/highway_assign.py [hunk 3]
```diff
@@ -144,16 +152,55 @@ def run(self):
...
+                if iteration == 0 or not run_dynamic_toll:
+                    assign_spec = self._get_assignment_spec(assign_classes)
+                    self._run_sola_traffic_assignment(scenario, assign_spec, chart_log_interval=1)
+                else: #iteration > 0 and run_dynamic_toll 
+                    # run maximum 5 times of dynamic tolling
+                    # break out the loop if no valuetoll need to be updated
+                    for dynamic_toll_iteration in range(1, 6):
+                        assign_spec = self._get_assignment_spec(assign_classes)
+                        if dynamic_toll_iteration < 5:
+                            assign_spec["stopping_criteria"]["max_iterations"] = self.controller.config.highway.tolls.dynamic_toll_inner_iter
```
Comment (i-am-sijia) [inline tm2py/components/network/highway/highway_assign.py]:
```suggestion
                            assign_spec["stopping_criteria"]["max_iterations"] = self.config.tolls.dynamic_toll_inner_iter
```

File: tm2py/components/network/highway/highway_assign.py [hunk 4]
```diff
@@ -144,16 +152,55 @@ def run(self):
...
+                        # if need to update valuetoll, call set_tolls function
+                        if update_toll_required:
+                            PrepareNetwork(controller=self._controller)._set_tolls(network=network, time_period=time)
+                            scenario.publish_network(network)
+                        else:
+                            if dynamic_toll_iteration < 5:
+                                # if there is no valuetoll updated needed before 5th dynamic toll iteration,
+                                # run remaining iterations upto the original max_iteration settings
+                                assign_spec = self._get_assignment_spec(assign_classes)
+                                assign_spec["stopping_criteria"]["max_iterations"] -= self.controller.config.highway.tolls.dynamic_toll_inner_iter
```
Comment (i-am-sijia) [inline tm2py/components/network/highway/highway_assign.py]:
```suggestion
                                assign_spec["stopping_criteria"]["max_iterations"] -= self.config.tolls.dynamic_toll_inner_iter
```
Can we really reduce the iteration here? Meaning does the assignment keep and restart from the previous state (with the inner iterations)?

File: tm2py/components/network/highway/highway_assign.py [hunk 5]
```diff
@@ -144,16 +152,55 @@ def run(self):
...
+                                        update_toll_required = True
+                                        link["@update_dynamic_toll"] = 1
+                                        break 
+
+                        # if need to update valuetoll, call set_tolls function
+                        if update_toll_required:
+                            PrepareNetwork(controller=self._controller)._set_tolls(network=network, time_period=time)
+                            scenario.publish_network(network)
+                        else:
+                            if dynamic_toll_iteration < 5:
```
Comment (i-am-sijia) [inline tm2py/components/network/highway/highway_assign.py:191]: This block of code about dynamic tolling makes the main function `run()` really long, suggest splitting it out into a `_run_dynamic_tolling_assignment()` function. This is not urgent to address, just esthetic.

**Review by i-am-sijia — CHANGES_REQUESTED** [review]
LGTM. And you have tested and checked the outputs, correct?

File: tm2py/components/network/highway/highway_assign.py (hunk 2 above)
Comment (ychtw) [reply tm2py/components/network/highway/highway_assign.py]: updated in 5eacf05

File: tm2py/components/network/highway/highway_assign.py (hunk 3 above)
Comment (ychtw) [reply tm2py/components/network/highway/highway_assign.py]: updated in 5eacf05

File: tm2py/components/network/highway/highway_assign.py (hunk 1 above)
Comment (ychtw) [reply tm2py/components/network/highway/highway_assign.py]: updated in 5eacf05

File: tm2py/components/network/highway/highway_assign.py (hunk 4 above)
Comment (ychtw) [reply tm2py/components/network/highway/highway_assign.py]: updated in d65a72d

**Review by i-am-sijia — APPROVED** [review]

**Comment by i-am-sijia** [comment]
Just approved the dynamic tolling changes. TO DO: @ychtw the MSA logic still needs to be refactored to reflect what we discussed today. Please create another PR for that.

**Pull request merged**